File: example/js/tmr/webcamera.js
'use strict';

// TMR web camera: enumerate the attached cameras, pick the first one that advertises a usable
// stream profile, open a session on it, and capture a single frame.
//
// This is the shortest complete path through the camera API, and the order is not negotiable:
// enumerate() names devices, capabilities() describes one of them, open() fixes the stream format
// for the whole session, start() begins delivery, and only then does readFrame() answer.
//
// Hardware-aware by design: a machine with no camera says so in one sentence and returns without
// failing, so the sample runs anywhere. A camera that enumerates but advertises no usable profile
// is a different matter and is raised as an error, matching the C++ counterpart's exit code.
// Build with WSE_BUILD_TMR=ON and WSE_BUILD_NODE_BINDING=ON.
//
// The argument is either the installed `lang/js` package directory or a built `wse.node` addon.
// Given the addon the sample still loads the package, so what is exercised is the public
// JavaScript surface an application uses rather than the raw native entry points.
//
//   node example/js/tmr/webcamera.js <installed lang/js directory or wse.node>
//
// This sample takes no arguments of its own.

const path = require('node:path');

const target = process.argv[2] || process.env.WSE_JAVASCRIPT_PACKAGE;
if ( !target ) {
  throw new Error('Pass the installed lang/js directory, or the built wse.node addon.');
}
// A `.node` argument is the build-tree addon. Point the package loader at it and still require the
// package, so the sample always demonstrates the same public surface an application uses.
if ( target.endsWith('.node') ) process.env.WSE_NODE_ADDON = path.resolve(target);
const wse = require(target.endsWith('.node')
  ? path.resolve(__dirname, '..', '..', '..', 'lang', 'js')
  : path.resolve(target));

async function main() {
  // enumerate() takes no argument here, which means CameraBackend.Automatic: every backend the
  // platform offers. Passing MediaFoundation, Video4Linux2, or Libcamera restricts the search to
  // that one. An empty list is a normal answer, not an error.
  const devices = wse.WebCamera.enumerate();
  if ( devices.length === 0 ) {
    console.log('No USB/UVC web camera is attached.');
    return;
  }

  // capabilities() is static and queries the device rather than a session, so the format can be
  // chosen before anything is opened. A device may enumerate and still advertise nothing usable,
  // so every candidate is tried instead of trusting the first.
  let device;
  let profile;
  for ( const candidate of devices ) {
    const capability = wse.WebCamera.capabilities(candidate);
    const candidateProfile = capability.streamProfiles[0];
    if ( candidateProfile && candidateProfile.outputFormats.length !== 0 ) {
      device = candidate;
      profile = candidateProfile;
      break;
    }
  }
  if ( !device ) {
    throw new Error('No enumerated camera has a usable stream profile.');
  }

  // The object owns the native device from open() until close(). JavaScript has no deterministic
  // destructor, so the try/finally is the whole release story: skip close() and the device stays
  // held, and unavailable to every other process, until the garbage collector finalizes this
  // object at some unpredictable later moment.
  const camera = new wse.WebCamera();
  try {
    // The configuration is fixed for the whole session. nativeFormat is what the device is asked
    // to produce, outputFormat what the caller wants back, and allowConversion lets TMR bridge
    // the two, for instance decoding MJPEG into BGR8. Set it false and a device that cannot
    // deliver outputFormat itself makes open() fail rather than paying for a conversion.
    camera.open(device, {
      nativeFormat: profile.nativeFormat,
      outputFormat: profile.outputFormats[0],
      allowConversion: true,
    });
    // start() with no argument is the pull form: frames are fetched by readFrame(). Passing a
    // function instead switches to the push form, where the session's worker thread delivers each
    // frame through the loop.
    camera.start();
    // readFrame() resolves a promise from a libuv worker thread, so the await yields rather than
    // blocking. The 2000 ms deadline is sized for a first frame, which is the slow one: a UVC
    // camera needs time to start its stream, and a much smaller value times out before the sensor
    // has delivered anything. The frame's `data` is a copy owned by JavaScript, so it stays valid
    // after stop() and close().
    const frame = await camera.readFrame(2000);
    console.log(`${frame.width}x${frame.height} bytes=${frame.data.length}`);
    camera.stop();
  } finally {
    // Reached on the ordinary path too, where streaming has already stopped; stop() is only
    // repeated when readFrame() threw. close() ends the session, is safe to call twice, and
    // leaves the object usable, so a later open() would start a new session on this same camera.
    if ( camera.isStreaming() ) camera.stop();
    camera.close();
  }
}

main().catch((error) => {
  console.error(error.category, error.code, error.nativeCode, error.message);
  process.exitCode = 1;
});

File: example/python/tmr/webcamera.py
"""Capture one owned frame through the portable WSE WebCamera API.

This is the shortest path from nothing to a frame: enumerate, pick a stream profile, open, start,
read one frame, and let the context manager put everything back. The other TMR samples build on
exactly this shape.

Unlike its neighbours this sample imports the installed `wse` package directly instead of taking
a module path on the command line, so it has to be run against an installed binding. It also
assumes a build that carries TMR; a build without it registers no `wse.WebCamera` at all.

Hardware-free by design in the ordinary case: a machine with no camera prints one line and returns
0, because an absent device is a fact about the machine and not a defect in the sample. Build with
WSE_BUILD_TMR=ON and WSE_BUILD_PYTHON_BINDING=ON.

The binding has no image writer and no log facility, so the frame is reported by shape and byte
count rather than saved, and the report goes through print.

    python example/python/tmr/webcamera.py
"""

import wse


def main() -> int:
    # enumerate() lists USB/UVC devices only; CSI, network, and virtual cameras are filtered out.
    # An empty list on a laptop that shows a virtual meeting camera is therefore the right answer.
    devices = wse.WebCamera.enumerate()
    if not devices:
        print("No USB/UVC web camera is attached.")
        return 0

    # capabilities() is the static form: it inspects a device it does not own, so no session
    # exists yet and there is nothing to close if it fails. Once a camera is open,
    # current_capabilities() is the one that describes the running session. A profile with no
    # output format cannot be streamed, so the first one that advertises one wins.
    device = None
    profile = None
    for candidate in devices:
        capability = wse.WebCamera.capabilities(candidate)
        if capability.stream_profiles and capability.stream_profiles[0].output_formats:
            device = candidate
            profile = capability.stream_profiles[0]
            break
    if device is None or profile is None:
        raise RuntimeError("No enumerated camera has a usable stream profile.")

    # The configuration separates what the sensor captures from what the caller wants to receive.
    # allow_conversion True lets TMR convert between them, which is what makes an MJPEG or YUYV
    # camera deliver the requested output; False demands the device produce output_format natively
    # and turns open() into a failure when it cannot.
    configuration = wse.CameraStreamConfiguration()
    configuration.native_format = profile.native_format
    configuration.output_format = profile.output_formats[0]
    configuration.allow_conversion = True

    # The context manager owns the camera: __exit__ calls release(), which stops the stream, closes
    # the session, and gives the native camera back for good. close() alone would end only the
    # session and leave the object open()able again, which is what resolution_change.py relies on.
    with wse.WebCamera() as camera:
        # open() fixes the resolution for the whole session and creates the session only; start()
        # is still needed before any frame arrives.
        camera.open(device, configuration)
        camera.start()
        try:
            # The timeout bounds the wait for this one frame. Two seconds is many frame periods at
            # any normal rate, so a healthy camera answers well inside it, while a value of a few
            # tens of milliseconds would turn an ordinarily slow first frame into a Timeout. On a
            # stopped session read_frame() reports NotStreaming immediately rather than waiting.
            frame = camera.read_frame(2000)
            print(f"{frame.width}x{frame.height} bytes={len(frame.data)}")
        finally:
            # stop() ends streaming and leaves the device open; the release() at the end of the
            # with block is what actually gives the camera back. The frame owns a copy of its
            # pixel bytes, so it stays readable after both.
            if camera.is_streaming:
                camera.stop()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
